### ISSUE TITLE
Setup Jenkins CI for building Pull Requests

### ISSUE THREAD

**Issue by daniilyar** [checkstyle/checkstyle#5953]
After discussion with @romani, we should setup the Jenkins server and give it an address *jenkins.checkstyle.org*.
This Jenkins instance should be publically accessible, so anybody who posted the new PR should be able to review the build log for it in Jenkins
Jenkins should build each pull request in this repository using the Jenkins pipeline job. Jenkins pipeline job code (aka Jenkinsfile) should be stored in this repository.
At the very first step, Jenkins build pipeline should just run the simple 'ls' command. This will allow us to not fail any PRs until the Jenkins PR testing workflow itself is well-tested.

After we test the PR building workflow well enough, we should create the separate issue to setup real tests in the pipeline, such as running UTs with Maven and so on.

Then we should use Hyper.sh as the main Jenkins computing resources provider (i.e. any CPU or memory-aggressive Jenkins build should happen in the Hyper.sh container).

Initially, Jenkins master can run builds on it's own executors, but eventually, we should have few Jenkins slaves running under the master, so master will only act as a simple job scheduler and UI while all builds will actually happen on slaves and delegated from slaves to Hyper.sh.

**Comment by daniilyar** [comment]
Installed Jenkins at the free Amazon AWS account: http://jenkins.checkstyle.org/

The AWS account credentilas and all other required credentials and info is synced directly with repo owner (@romani).

I have configured the GitHub auth on Jenkins, so every Checkstyle organization member now have an access to edit Jenkins configs.

I have also created the job for building pull requests with the Jenkins pipeline stored in git - http://jenkins.checkstyle.org/job/pr-builder-checkstyle-pipeline/

The name of job and it's config is a subject to change in futere, at the current step my goal is to setup a working proof-of-concept or building PRs at Jenkins with a coded pipeline.

@rnveach , you had a server which we may use as a Jenkins slave, right? Please contact me so we can proceed with configuring it into a Jenkins slave.

Taking into account the fact that we are provided with some computing resources by Hyper.sh, the end goal is to setup the following workflow:
1. Jenkins see the pull requests and starts to build it at the Jenkins slave.
2. Build triggers many parallel jobs, such as running MAven UTs, each running in the separate Hyper.sh container
3. Jenkins accumulates test results and posts the PR build status on the PR in the same way as it is done by the other Checkstyle CIs such as Travis and Wercker.

**Comment by romani** [comment]
@daniilyar , please plan to move to Jenkins, following configs from Travis.
`No error testing - simple-binary-encoding`
`NoExceptiontest - Guava with google_checks (oraclejdk8)`
`Ensure that all Sevntu check are kused`
`Check the chmod on files.`
`https://sonarcloud.io (oraclejdk8)`

Sonar will require secret token, so it will be good point to investigate how to make it by Jenkins.

**Comment by daniilyar** [comment]
@rnveach thank you for the server,I have connected it to our Jenkins as a Jenkins slave, but currently there is a 'permission denied' error when pulling public (!) Checkstyle repo. Could you please help me to fix that properly from your side?

**Comment by pbludov** [comment]
@daniilyar, did you properly configure the ssh connection? Here is the guide:
https://help.github.com/articles/connecting-to-github-with-ssh/

**Comment by daniilyar** [comment]
@pbludov , I have found an issue with SSH key on the lave and fixed it, so now it works. So, now every build happens on Jenkins slave and master is now staying for UI access only.

**Comment by daniilyar** [comment]
@romani , @rnveach, starting from today, all Checkstyle PRs are now additionally tested by Jenkins.
For now, each PR Jenkins will run 2 checks - 'mvn compile' and 'mvn package' in parallel in 2 Separate Hyper.sh containers. In the meanwhile I will proceed with duplicating existing Travis builds on Jenkins to test Jenkins under the real CI load and prove it's stability.

**Comment by daniilyar** [comment]
> Sonar will require secret token, so it will be good point to investigate how to make it by Jenkins.
@romani keeping secrets on Jenkins will require us to setup HTTPS for it. So for Jenkins I am going to setup the auto-updating Lets Encrypt SSL cert.

**Comment by romani** [comment]
@daniilyar ,

1)
> could you please review the Jenkins job config and this example build log?

after `[Compile] [pr-builder-checkstyle] Running shell script` please print command that user can run on his local to reproduce `mvn -B compile`.

2)
> [Package] [INFO] Downloaded from central

downloads become visible , is it a time to make new docker image ?

3) it is very helpful to have numerated lines in web logs, to point contributor to certain line.
Example: https://travis-ci.org/checkstyle/checkstyle/jobs/400679060#L813
Contributors will never read logs attentively, so we point them to root of problems all the time.
I am not sure if jenkins have such plugin.

**Comment by daniilyar** [comment]
@romani thank you for review, will address all your points soon! In the meanwhile, I enabled the SSL for Jenkins with auto-renewing Let'sEncrypt cert: https://jenkins.checkstyle.org/

**Comment by daniilyar** [comment]
> downloads become visible , is it a time to make new docker image ?
@romani, I've just built the new docker image. Also added Jenkins job that allows to rebuild the docker image in a single click: https://jenkins.checkstyle.org/job/docker-maven-build-image/

**Comment by dcsobral** [comment]
Was that discussion that led to this anywhere publicly visible? Could you paste a link to it if so, or mention the reasons if not?

**Comment by romani** [comment]
Our dockerhub organization images are at https://hub.docker.com/u/checkstyle/ but please be sure that such images are for our CI , so we can change them in any way we might need.

**Comment by daniilyar** [comment]
Jenkins is finally configured to build all PRs, so I am closing this ticket

**Comment by daniilyar** [comment]
Project Jenkins is finally configured to build all PRs from checkstyle/checkstyle repository, so I am closing this ticket
